### PR TITLE
sql: don't use the grammar to enforce expression const-ness in AOST

### LINKED ISSUE #26976: sql: show syntax with AOST interval broken?
Ok maybe this one is a real problem?:

```
> show syntax 'select * from t as of system time interval ''-1ns''';
+-------+-----------------------------------------------------+
| field |                       message                       |
+-------+-----------------------------------------------------+
| sql   | SELECT * FROM t AS OF SYSTEM TIME '-1ns':::INTERVAL |
+-------+-----------------------------------------------------+

> SELECT * FROM t AS OF SYSTEM TIME '-1ns':::INTERVAL;
invalid syntax: statement ignored: syntax error at or near ":"
```

The FmtParsable flag that is used here by `show synatx` moves the interval type annotation to after its argument and adds on a `:::<type>`. However the AOST parser doesn't accept that syntax and so it breaks. However I have no idea if there are any bugs here.

1) Is this even a bug?
2) Should FmtParsable not move those type annotations like it is? Can we do something like teach tree.AnnotateTypeExpr about this?
3) Should AOST be taught about type annotations? I think this is fairly difficult because they aren't constants. AOST is clearly being a bit special here. But even still we should probably have a rule that show syntax's output works exactly the same as its input.

### PR DESCRIPTION
**Pull request by knz**
Fixes  #26976.

Prior to this patch, AS OF SYSTEM TIME relied on the grammar to
restrict its operand to just constants. This was an unfortunate
approach because it prevented other constant expressions like `('a' ||
'b')::INTERVAL` (or, encountered during tests, `'1ms':::INTERVAL`).

This patch corrects this situation by relaxing the grammar constraint
and instead verifying the const-ness of the expression after type
checking.

Release note (sql change): AS OF SYSTEM TIME can now use some more
complex expressions to compute the desired time stamp.

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/as_of [hunk 1]
```diff
@@ -23,6 +23,15 @@ SELECT * FROM t AS OF SYSTEM TIME INTERVAL '-1ns'
 ----
 2
 
+# Verify that we can use computed expressions.
```
Comment (mjibson) [inline pkg/sql/logictest/testdata/logic_test/as_of]: Add another test with `AOST now()` to verify non-constant expressions don't work.

**Comment by mjibson** [comment]
Also could you remove the restriction in https://github.com/cockroachdb/cockroach/blob/ca89c4f3dee2631000c20aa7e110541ea88e19cc/pkg/testutils/sqlutils/pretty.go#L30 to make sure this fixes the pretty print round trip tests?

**Review by knz — COMMENTED** [review]
> Also could you remove the restriction in cockroach/pkg/testutils/sqlutils/pretty.go Line 30 in ca89c4f

Done

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27206#-:-LGxujz61t6ZjWJwHb0i:b-23ocpn)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/logictest/testdata/logic_test/as_of, line 26 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/27206#-LGe-ml76GcQ1pHgG-mJ-r1-26:-LGxunqB8uXyKDoCtuHN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b2f3f9da2338cbfecc8d8f1c397ae4820cdfa994/pkg/sql/logictest/testdata/logic_test/as_of#L26)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Add another test with `AOST now()` to verify non-constant expressions don't work.
</blockquote></details>

Done.

**Review by madelynnblue — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/27206#-:-LH-Hm6c1LlK50U18ylc:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Comment by knz** [comment]
Thank you!

bors r+
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Pull request merged**